solr/core/src/java/org/apache/solr/util/EnvUtils.java

```diff
@@ -0,0 +1,279 @@
+    return getPropAsLong(key, null);
+  }
+
+  /** Get property as long, or default value */
+  public static Long getPropAsLong(String key, Long defaultValue) {
+    String value = getProp(key);
+    if (value == null) {
+      return defaultValue;
+    }
+    return Long.parseLong(value);
+  }
+
+  /** Get property as boolean */
+  public static Boolean getPropAsBool(String key) {
+    return getPropAsBool(key, null);
+  }
+
+  /** Get property as boolean, or default value */
+  public static Boolean getPropAsBool(String key, Boolean defaultValue) {
+    String value = getProp(key);
+    if (value == null) {
+      return defaultValue;
+    }
+    return StrUtils.parseBool(value);
+  }
+
+  /**
+   * Get comma separated strings from sysprop as List
+   *
+   * @return list of strings, or null if not found
+   */
+  public static List<String> getPropAsList(String key) {
+    return getPropAsList(key, null);
+  }
+
+  /**
+   * Get comma separated strings from sysprop as List, or default value
+   *
+   * @return list of strings, or provided default if not found
+   */
+  public static List<String> getPropAsList(String key, List<String> defaultValue) {
+    return getProp(key) != null ? stringValueToList(getProp(key)) : defaultValue;
+  }
+
+  /** Set a system property. Shim to {@link System#setProperty(String, String)} */
+  public static void setProp(String key, String value) {
+    System.setProperty(key, value);
+    System.setProperty(camelCaseToDotSeparated(key), value);
+  }
+
+  /**
+   * Re-reads environment variables and updates the internal map.
+   *
+   * @param overwrite if true, overwrite existing system properties with environment variables
+   */
+  public static synchronized void init(boolean overwrite) {
+    // Convert eligible environment variables to system properties
+    for (String key : ENV.keySet().toArray(String[]::new)) {
+      if (key.startsWith("SOLR_") || CUSTOM_MAPPINGS.containsKey(key)) {
+        String sysPropKey = envNameToSyspropName(key);
+        // Existing system properties take precedence
+        if (!sysPropKey.isBlank() && (overwrite || getProp(sysPropKey) == null)) {
+          setProp(sysPropKey, ENV.get(key));
+        }
+      }
+    }
+    initialized = true;
+  }
+
+  protected static String envNameToSyspropName(String envName) {
+    return CUSTOM_MAPPINGS.containsKey(envName)
+        ? CUSTOM_MAPPINGS.get(envName)
+        : envName.toLowerCase(Locale.ROOT).replace("_", ".");
+  }
+
+  @SuppressWarnings("unchecked")
+  private static List<String> stringValueToList(String string) {
+    if (string.startsWith("[") && string.endsWith("]")) {
+      // Convert a JSON string to a List<String> using Noggit parser
+      return (List<String>) Utils.fromJSONString(string);
+    } else {
+      return StrUtils.splitSmart(string, ",", true).stream()
+          .map(String::trim)
+          .collect(Collectors.toList());
+    }
+  }
+
+  private static synchronized void ensureInitialized() {
```

Comment:
Since we're working with static global methods and state here, the initialization order is not well defined (unless we use some IOC stuff). Say you have a class declaring `public static final String MY_VAR = EnvUtils.getProp("solr.foo")`. Now, if user sets `SOLR_FOO` as env.var, then we need to make sure `EnvUtils.init()` has already run.

There may be better ways to achieve this than busy-waiting, ideas?

Reply:
Hmm, this bit me, cause I call `getProp` from within `init()` itself, and then we got a deadlock so Solr would not start at all. So it is a bit fragile, we must make sure there is no circular calls with the same lock, or find another safe way for late or async initialization..

Reply:
I'm not quite following.  A static initializer is guaranteed to run to completion before any other threads may proceed to call into it (they block).  And you have a static initializer calling init(false).  init(true) is only used by some tests; we can ignore synchronization for them.

Reply:
For internal use in this one class that's ok. But when another class calls one of the `getProp` methods that require initialization, we have to block.

Reply:
We agree blocking is necessary.  I claim the way a JVM loads classes with a static initializer will accomplish that without code specific in this class to try to accomplish similarly.

Reply:
The JVM does not know about our `init()` method, and that e.g. `getProp()` depends on `init()` having ran. We could expclicitly load this `EnvUtils` class early by calling it  at the earliest possible time, and JVM would make sure it was initialized before others. But locking is safer.

Reply:
The JVM will block any thread from calling `getProp` (or accessing `EnvUtils` in any way) until `EnvUtils` is initialized.  Initialization of this class calls init().  See [section 12.4.2 of the Java Language Specification](https://docs.oracle.com/javase/specs/jls/se8/html/jls-12.html#jls-12.4.2), particularly references to "LC".  

I'm aware init() indirectly calls getProp but it proceeds because this thread holds the LC lock.

Reply:
Hum, I’m sure it did not work before adding the lock, but I see what the spec claims here, so I’ll have to give it a try by removing locks.. Would indeed simplify all.